File: packages/engine/src/core/index-pass.ts
import { join } from 'node:path'
import type { CorrectionPayload } from '@sofar/schema'
import {
  DEFAULT_META_FILE,
  INDEX_SCHEMA_VERSION,
  readIndexMeta,
  writeIndexMeta,
  type IndexMeta,
  type InitiativeCursor,
} from './index-store'
import { readSince, type IndexedEvent } from './index-tail'
import { initiativeSlugs } from './listing'

/**
 * One incremental pass over every log, shared by every tier (record-index 3.1).
 *
 * Tier 0 shipped with this loop inline. Tier 1 needs the same one, and copying
 * it would have copied the parts that are easy to get wrong rather than the
 * part that is easy to see — so the loop lives here once and a tier supplies
 * only its reducer: what an empty state is, and what one event does to it.
 *
 * Everything hard is in this file, and all of it is about the same question:
 * WHEN IS RESUMING UNSOUND? A cursor makes the common case cheap; these are
 * the four cases where the cheap answer would be the wrong one, each falling
 * back to reading the whole log rather than to a guess.
 *
 * 1. THE CURSOR DOES NOT DESCRIBE THIS FILE — an import, a rewrite, a restore.
 *    Detected by the tail reader's own corroboration (index-tail.ts).
 * 2. THE TAIL IS NOT IN ORDER. The fold replays in ULID order, not file order
 *    (the convergent-fold rule), and `merge=union` on events.jsonl interleaves
 *    two branches' lines — the merge strategy this project chose. Applying
 *    those incrementally would order them differently from every other surface,
 *    so an id that is not above everything already applied forces a rebuild,
 *    and a rebuild sorts by id exactly as the fold does.
 * 3. A CORRECTION VOIDS SOMETHING ALREADY APPLIED. This is the one retroactive
 *    act in an append-only log: `correction` names an event id, and the fold
 *    drops that event wherever it sits. A pass that ignores it keeps reporting
 *    work the record has withdrawn — a session that never really started, a
 *    decision that was taken back. Since applied state cannot be un-applied,
 *    a correction reaching back past the batch rebuilds the log. This record
 *    carries 24 of them, over session_started, session_ended, decision_logged
 *    and command_run, so it is a live case rather than a theoretical one.
 * 4. A LATER EVENT WAS ALREADY VOIDED. Voids are therefore remembered on the
 *    cursor, not just applied to the batch that carried them.
 *
 * The cost of being wrong here is silent and permanent, and the cost of being
 * conservative is one re-read of one log. That is the whole trade.
 */

/** What a tier must supply: its per-slug state and how one event changes it. */
export interface SlugReducer<S> {
  /** A fresh state — cold start, or any of the rebuild cases above. */
  empty: () => S
  /** Copy state read from disk, so a resume never mutates it in place. */
  clone: (state: S) => S
  /** Apply one event, exactly as the fold would. */
  apply: (state: S, event: IndexedEvent, slug: string) => void
}

export interface PassResult<S> {
  /** Per-slug state, for every initiative that exists right now. */
  states: Record<string, S>
  /** False when nothing moved — callers skip writing what has not changed. */
  changed: boolean
}

/** The `ref` of a correction, or null when the payload does not carry one. */
function voidedRef(event: IndexedEvent): string | null {
  if (event.type !== 'correction') return null
  const ref = (event.payload as unknown as CorrectionPayload).ref
  return typeof ref === 'string' && ref.length > 0 ? ref : null
}

/** Ids strictly ascending, and all above what has already been applied. */
function inOrder(events: readonly IndexedEvent[], after: string | undefined): boolean {
  let previous = after
  for (const event of events) {
    if (previous !== undefined && event.id <= previous) return false
    previous = event.id
  }
  return true
}

/** Greatest id in a batch, or the prior maximum when the batch is empty. */
function maxIdOf(events: readonly IndexedEvent[], prior: string | undefined): string | undefined {
  let max = prior
  for (const event of events) if (max === undefined || event.id > max) max = event.id
  return max
}

export function passOverRecord<S>(
  sofarDir: string,
  metaFile: string,
  prior: Record<string, S> | null,
  reducer: SlugReducer<S>,
): PassResult<S> {
  const meta: IndexMeta = readIndexMeta(sofarDir, metaFile) ?? {
    version: INDEX_SCHEMA_VERSION,
    cursors: {},
  }
  const states: Record<string, S> = {}
  let changed = prior === null

  for (const slug of initiativeSlugs(sofarDir)) {
    const log = join(sofarDir, 'initiatives', slug, 'events.jsonl')
    const cursor = meta.cursors[slug] ?? null
    let read = readSince(log, cursor)
    let voided = new Set(cursor?.voided ?? [])

    // Case 2 and case 3, both of which mean "what I have is not a prefix of
    // what is true". Checked BEFORE anything is applied, so the fallback costs
    // one extra read rather than a corrupted state.
    if (!read.full && read.events.length > 0) {
      const batch = new Set(read.events.map((e) => e.id))
      const reachesBack = read.events.some((e) => {
        const ref = voidedRef(e)
        return ref !== null && !batch.has(ref)
      })
      if (reachesBack || !inOrder(read.events, cursor?.maxId ?? cursor?.id)) {
        read = readSince(log, null)
      }
    }

    const rebuilt = read.full || prior === null || prior[slug] === undefined
    if (read.full) voided = new Set() // a rebuild re-derives every void below

    // Corrections first, exactly as the fold's pre-pass does: a correction
    // voids its target wherever that target sits, including later in the same
    // batch. Then replay in ulid order — file order is not normative.
    const events = read.full
      ? [...read.events].sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0))
      : read.events

    // A quiet initiative is carried forward BY REFERENCE. `clone` exists so a
    // resume never mutates prior state in place, and with nothing to apply
    // there is nothing to mutate — the copy would be written back byte for
    // byte. It is the difference between O(new events) and O(repo) per
    // refresh: at 1000 initiatives one appended event was deep-copying 50,272
    // path entries, which is most of what the derived half cost to read
    // (record-index 3.3). Safe because `prior` is a fresh parse owned by this
    // call and the states it returns are serialized, never mutated.
    const state = rebuilt
      ? reducer.empty()
      : events.length === 0
        ? prior[slug]!
        : reducer.clone(prior[slug]!)
    for (const event of events) {
      const ref = voidedRef(event)
      if (ref !== null) voided.add(ref)
    }
    for (const event of events) {
      if (voided.has(event.id)) continue
      reducer.apply(state, event, slug)
    }

    states[slug] = state
    // A rebuild that read nothing (an absent or unreadable log) lands on the
    // same empty state it had, so it is not a change — otherwise a record with
    // one logless initiative would rewrite every index file on every pass.
    if (events.length > 0 || prior === null || prior[slug] === undefined) changed = true
    else if (read.full && read.cursor !== null) changed = true

    if (read.cursor === null) {
      if (meta.cursors[slug] !== undefined) {
        delete meta.cursors[slug]
        changed = true
      }
    } else {
      const next: InitiativeCursor = { ...read.cursor }
      const max = maxIdOf(events, read.full ? undefined : cursor?.maxId)
      if (max !== undefined) next.maxId = max
      if (voided.size > 0) next.voided = [...voided].sort()
      meta.cursors[slug] = next
    }
  }

  // An initiative that vanished must not linger in the cursor map.
  for (const slug of Object.keys(meta.cursors)) {
    if (states[slug] === undefined) {
      delete meta.cursors[slug]
      changed = true
    }
  }
  if (prior !== null && Object.keys(prior).some((slug) => states[slug] === undefined)) changed = true

  if (changed) writeIndexMeta(sofarDir, meta, metaFile)
  return { states, changed }
}

export { DEFAULT_META_FILE }

File: crates/sofar-core/src/index_pass.rs
//! One incremental pass over every log, shared by every tier
//! (`core/index-pass.ts`, record-index 3.1): a tier supplies its reducer;
//! the five "resuming is unsound" cases here fall back to a full read.

use crate::index_store::{Cursor, IndexMeta, read_index_meta, write_index_meta};
use crate::index_tail::{IndexedEvent, read_since};
use crate::json::Json;
use crate::layout::{Layout, initiative_slugs};
use crate::text::cmp_utf16;

/// What a tier must supply.
pub trait SlugReducer {
    type State: Clone;
    fn empty(&self) -> Self::State;
    fn apply(&self, state: &mut Self::State, event: &IndexedEvent, slug: &str);
}

#[derive(Debug)]
pub struct PassResult<S> {
    /// Per-slug state for every initiative that exists right now, in slug order.
    pub states: Vec<(String, S)>,
    /// False when nothing moved.
    pub changed: bool,
}

fn voided_ref(event: &IndexedEvent) -> Option<&str> {
    if event.event_type != "correction" {
        return None;
    }
    event.payload.get("ref").and_then(Json::as_nonempty_str)
}

fn in_order(events: &[IndexedEvent], after: Option<&str>) -> bool {
    let mut previous = after.map(str::to_owned);
    for event in events {
        if let Some(p) = &previous
            && cmp_utf16(&event.id, p).is_le()
        {
            return false;
        }
        previous = Some(event.id.clone());
    }
    true
}

fn max_id_of(events: &[IndexedEvent], prior: Option<&str>) -> Option<String> {
    let mut max = prior.map(str::to_owned);
    for event in events {
        if max
            .as_deref()
            .is_none_or(|m| cmp_utf16(&event.id, m).is_gt())
        {
            max = Some(event.id.clone());
        }
    }
    max
}

/// `passOverRecord`.
#[allow(clippy::too_many_lines, reason = "one loop, ported as written")]
pub fn pass_over_record<R: SlugReducer>(
    layout: &Layout,
    meta_file: &str,
    prior: Option<&[(String, R::State)]>,
    reducer: &R,
) -> PassResult<R::State> {
    let mut meta = read_index_meta(layout, meta_file).unwrap_or_default();
    let mut states: Vec<(String, R::State)> = Vec::new();
    let mut changed = prior.is_none();

    for slug in initiative_slugs(layout) {
        let log = layout.events_path(&slug);
        let prior_state = prior.and_then(|p| p.iter().find(|(s, _)| *s == slug).map(|(_, st)| st));
        let cursor: Option<Cursor> = if prior_state.is_none() {
            None
        } else {
            meta.get(&slug).cloned()
        };
        let mut read = read_since(&log, cursor.as_ref());
        let mut voided: Vec<String> = cursor
            .as_ref()
            .and_then(|c| c.voided.clone())
            .unwrap_or_default();

        if !read.full && !read.events.is_empty() {
            let reaches_back = read
                .events
                .iter()
                .any(|e| voided_ref(e).is_some_and(|r| !read.events.iter().any(|b| b.id == r)));
            let after = cursor
                .as_ref()
                .and_then(|c| c.max_id.as_deref().or(Some(c.id.as_str())));
            if reaches_back || !in_order(&read.events, after) {
                read = read_since(&log, None);
            }
        }

        let rebuilt = read.full || prior_state.is_none();
        if read.full {
            voided.clear();
        }
        let mut events = read.events;
        if read.full {
            events.sort_by(|a, b| cmp_utf16(&a.id, &b.id));
        }

        let mut state = if rebuilt {
            reducer.empty()
        } else {
            prior_state.expect("not rebuilt").clone()
        };
        for event in &events {
            if let Some(r) = voided_ref(event)
                && !voided.iter().any(|v| v == r)
            {
                voided.push(r.to_owned());
            }
        }
        for event in &events {
            if voided.contains(&event.id) {
                continue;
            }
            reducer.apply(&mut state, event, &slug);
        }

        if !events.is_empty() || prior_state.is_none() || (read.full && read.cursor.is_some()) {
            changed = true;
        }

        match read.cursor {
            None => {
                if meta.remove(&slug) {
                    changed = true;
                }
            }
            Some(mut next) => {
                let prior_max = if read.full {
                    None
                } else {
                    cursor.as_ref().and_then(|c| c.max_id.as_deref())
                };
                next.max_id = max_id_of(&events, prior_max);
                if !voided.is_empty() {
                    let mut sorted = voided.clone();
                    sorted.sort_by(|a, b| cmp_utf16(a, b));
                    next.voided = Some(sorted);
                }
                meta.set(&slug, next);
            }
        }
        states.push((slug, state));
    }

    let stale: Vec<String> = meta
        .cursors
        .iter()
        .filter(|(s, _)| !states.iter().any(|(t, _)| t == s))
        .map(|(s, _)| s.clone())
        .collect();
    for slug in stale {
        meta.remove(&slug);
        changed = true;
    }
    if let Some(prior) = prior
        && prior
            .iter()
            .any(|(s, _)| !states.iter().any(|(t, _)| t == s))
    {
        changed = true;
    }
    if changed {
        write_index_meta(layout, &meta, meta_file);
    }
    PassResult { states, changed }
}

/// The meta cursors as they stand, for tests and diagnostics.
#[must_use]
pub fn current_meta(layout: &Layout, meta_file: &str) -> Option<IndexMeta> {
    read_index_meta(layout, meta_file)
}
